src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/Delegation/DesignTimeHelperResponseRewriter.cs

```diff
@@ -59,57 +55,49 @@ internal class DesignTimeHelperResponseRewriter : DelegatedCompletionResponseRew
 
         var sourceText = await hostDocumentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
 
-        // We should remove Razor design time helpers from C#'s completion list. If the current identifier being targeted does not start with a double
-        // underscore, we trim out all items starting with "__" from the completion list. If the current identifier does start with a double underscore
-        // (e.g. "__ab[||]"), we only trim out common design time helpers from the completion list.
+        // We should remove Razor design-time helpers from C#'s completion list. If the current identifier
+        // being targeted does not start with a double underscore, we trim out all items starting with "__"
+        // from the completion list. If the current identifier does start with a double underscore (e.g. "__ab[||]"),
+        // we only trim out common design time helpers from the completion list.
 
-        var filteredItems = completionList.Items.Except(s_designTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
+        using var _ = ListPool<CompletionItem>.GetPooledObject(out var filteredItems);
 
-        if (ShouldRemoveAllDesignTimeItems(owner, sourceText))
-        {
-            filteredItems = filteredItems.Where(item => item.Label != null && !item.Label.StartsWith("__", StringComparison.Ordinal)).ToArray();
-        }
+        var items = completionList.Items;
+        filteredItems.SetCapacityIfLarger(items.Length);
 
-        completionList.Items = filteredItems;
-        return completionList;
-    }
+        var shouldRemoveAllDesignTimeItems = StartsWithDoubleUnderscore(owner, sourceText);
```

Comment:
The method name does not match the logic that is happening (and the comment above). We remove all "design time items" if the current identifier _doesn't_ start with a double underscore.

Also I think the term "design time items" here is weird, and we can just be super explicit: `var shouldRemoveAllDoubleUnderscoreItems = DoesNotStartWithDoubleUnderscore(owner, sourceText);`

Reply:
Though given the tests all pass, maybe I'm reading something wrong here?

Reply:
I'll push that comment closer to where it's used and flip the method to be positive rather than negative. I hope that'll help.

Reply:
Thank you, the updates make things much clearer